// @datorama/akita/esm2015/src/plugins/plugin.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { filterNil } from '../api/operators';
import { toBoolean } from '../internal/utils';
import { getAkitaConfig } from '../api/config';
/**
 * @abstract
 * @template E, S
 */
export class AkitaPlugin {
    /**
     * @param {?} query
     * @param {?=} config
     */
    constructor(query, config) {
        this.query = query;
        if (config && config.resetFn) {
            if (getAkitaConfig().resettable) {
                this.onReset(config.resetFn);
            }
        }
    }
    /**
     * This method is responsible for getting access to the query.
     * @return {?}
     */
    getQuery() {
        return this.query;
    }
    /**
     * This method is responsible for getting access to the store.
     * @return {?}
     */
    getStore() {
        return this.getQuery().__store__;
    }
    /**
     * This method is responsible tells whether the plugin is entityBased or not.
     * @param {?} entityId
     * @return {?}
     */
    isEntityBased(entityId) {
        return toBoolean(entityId);
    }
    /**
     * This method is responsible for selecting the source; it can be the whole store or one entity.
     * @param {?} entityId
     * @return {?}
     */
    selectSource(entityId) {
        if (this.isEntityBased(entityId)) {
            return ((/** @type {?} */ (this.getQuery()))).selectEntity(entityId).pipe(filterNil);
        }
        return ((/** @type {?} */ (this.getQuery()))).select(state => state);
    }
    /**
     * @param {?} entityId
     * @return {?}
     */
    getSource(entityId) {
        if (this.isEntityBased(entityId)) {
            return ((/** @type {?} */ (this.getQuery()))).getEntity(entityId);
        }
        return this.getQuery().getSnapshot();
    }
    /**
     * This method is responsible for updating the store or one entity; it can be the whole store or one entity.
     * @param {?} newState
     * @param {?=} entityId
     * @return {?}
     */
    updateStore(newState, entityId) {
        if (this.isEntityBased(entityId)) {
            this.getStore().update(entityId, newState);
        }
        else {
            this.getStore().setState(state => (Object.assign({}, state, newState)));
        }
    }
    /**
     * Function to invoke upon reset
     * @param {?} fn
     * @return {?}
     */
    onReset(fn) {
        /** @type {?} */
        const original = this.getStore().reset;
        this.getStore().reset = (...params) => {
            /** It should run after the plugin destroy method */
            setTimeout(() => {
                original.apply(this.getStore(), params);
                fn();
            });
        };
    }
}
if (false) {
    /** @type {?} */
    AkitaPlugin.prototype.query;
    /**
     * This method is responsible for cleaning.
     * @abstract
     * @return {?}
     */
    AkitaPlugin.prototype.destroy = function () { };
}
