// @datorama/akita/esm2015/src/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
export {} from './api/types';
export { EntityStore, getInitialEntitiesState, getInitialActiveState } from './api/entity-store';
export { QueryEntity } from './api/query-entity';
export { Query } from './api/query';
export { enableAkitaProdMode, isDev, __stores__, rootDispatcher, Store } from './api/store';
export { applyTransaction, transaction } from './api/transaction';
export { push, remove, pop, unshift, sort, reverse, swap, update, splice, toggle } from './api/arrays';
export { filterNil } from './api/operators';
export { coerceArray, isObject, isPlainObject, isFunction, toBoolean, isUndefined, isNil, entityExists, noop, isObservable, isString, isNumber, isDefined, resetActive, isActiveState, setValue, getValue } from './internal/utils';
export { applyAction, action } from './internal/action';
export { mapInWorker } from './api/worker';
export { memoizeOne } from './api/memoize';
export { SnapshotManager, snapshotManager } from './api/snapshot-manager';
export { StoreConfig, configKey } from './api/store-config';
export { QueryConfig, queryConfigKey } from './api/query-config';
export { akitaConfig, getAkitaConfig } from './api/config';
export { compareValues, Order } from './internal/sort';
export { AkitaPlugin } from './plugins/plugin';
export { PaginatorPlugin, Paginator } from './plugins/paginator/paginator-plugin';
export { PersistNgFormPlugin } from './plugins/persist-form/persist-ng-form-plugin';
export { persistState } from './enhancers/persist-state';
export { akitaDevtools } from './enhancers/devtools';
export { EntityCollectionPlugin } from './plugins/entity-collection-plugin';
export { StateHistoryPlugin } from './plugins/state-history/state-history-plugin';
export { EntityStateHistoryPlugin } from './plugins/state-history/entity-state-history-plugin';
export { getNestedPath, dirtyCheckDefaultParams, DirtyCheckPlugin } from './plugins/dirty-check/dirty-check-plugin';
export { EntityDirtyCheckPlugin } from './plugins/dirty-check/entity-dirty-check-plugin';
export {} from './api/operators';
export { increment, decrement, guid, resetStores } from './api/store-utils';
export { __globalState } from './internal/global-state';
