// @datorama/akita/esm2015/src/internal/error.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { entityExists, isObject, toBoolean } from './utils';
export class AkitaError extends Error {
    /**
     * @param {?} message
     */
    constructor(message) {
        super(message);
    }
}
export class AkitaImmutabilityError extends AkitaError {
    /**
     * @param {?} storeName
     */
    constructor(storeName) {
        super(`The new state should be immutable. Make sure to return a new immutable state. \n store: \n ${storeName}`);
    }
}
export class AkitaEntityNotExistsError extends AkitaError {
    /**
     * @param {?} id
     */
    constructor(id) {
        super(`Entity ${id} does not exists`);
    }
}
export class AkitaNoActiveError extends AkitaError {
    constructor() {
        super(`Active is null/undefined`);
    }
}
export class AkitaInvalidEntityState extends AkitaError {
    constructor() {
        super(`Entity state is invalid`);
    }
}
export class AkitaUpdateIdKeyError extends AkitaError {
    constructor() {
        super(`Updating entity id is not permitted when updating many entities`);
    }
}
/**
 * @param {?} id
 * @param {?} entities
 * @return {?}
 */
export function assertEntityExists(id, entities) {
    if (!entityExists(id, entities)) {
        throw new AkitaEntityNotExistsError(id);
    }
}
/**
 * @param {?} store
 * @return {?}
 */
export function assertActive(store) {
    if (!toBoolean(store.active)) {
        throw new AkitaNoActiveError();
    }
}
/**
 * @param {?} state
 * @return {?}
 */
export function assertEntityState(state) {
    /** @type {?} */
    const assertEntities = isObject(state);
    if (!assertEntities) {
        throw new AkitaInvalidEntityState();
    }
}
/**
 * @param {?} name
 * @param {?} className
 * @return {?}
 */
export function assertDecorator(name, className) {
    if (!name) {
        console.error(`@StoreConfig({ name }) is missing in ${className}`);
    }
}
