// @datorama/akita/esm5/src/api/query.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { take } from 'rxjs/operators';
import { queryConfigKey } from './query-config';
/**
 * @template S
 */
var /**
 * @template S
 */
Query = /** @class */ (function () {
    function Query(store) {
        this.store = store;
        this.__store__ = store;
    }
    /**
     * @template R
     * @param {?=} project
     * @return {?}
     */
    Query.prototype.select = /**
     * @template R
     * @param {?=} project
     * @return {?}
     */
    function (project) {
        /** @type {?} */
        var state = project ? project : function (state) { return state; };
        return this.store._select(state);
    };
    /**
     * Select once and complete.
     */
    /**
     * Select once and complete.
     * @template R
     * @param {?} project
     * @return {?}
     */
    Query.prototype.selectOnce = /**
     * Select once and complete.
     * @template R
     * @param {?} project
     * @return {?}
     */
    function (project) {
        return this.select(project).pipe(take(1));
    };
    /**
     * Select the store's loading state.
     */
    /**
     * Select the store's loading state.
     * @return {?}
     */
    Query.prototype.selectLoading = /**
     * Select the store's loading state.
     * @return {?}
     */
    function () {
        return this.select(function (state) { return ((/** @type {?} */ (state))).loading; });
    };
    /**
     * Select the store's error state.
     */
    /**
     * Select the store's error state.
     * @return {?}
     */
    Query.prototype.selectError = /**
     * Select the store's error state.
     * @return {?}
     */
    function () {
        return this.select(function (state) { return ((/** @type {?} */ (state))).error; });
    };
    /**
     * Get the raw value of the store.
     */
    /**
     * Get the raw value of the store.
     * @return {?}
     */
    Query.prototype.getSnapshot = /**
     * Get the raw value of the store.
     * @return {?}
     */
    function () {
        return this.store._value();
    };
    Object.defineProperty(Query.prototype, "isPristine", {
        /**
         *  Returns whether the state of the store is pristine (the set() method hasn't been explicitly called
         *  since the store creation, or the last time setPristine()was called).
         */
        get: /**
         *  Returns whether the state of the store is pristine (the set() method hasn't been explicitly called
         *  since the store creation, or the last time setPristine()was called).
         * @return {?}
         */
        function () {
            return this.store.isPristine;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Query.prototype, "isDirty", {
        /**
         * Returns whether the state of the store is dirty (the set() method
         * or the setDirty() method has been explicitly called since the store creation,
         * or the last time setPristine() was called).
         */
        get: /**
         * Returns whether the state of the store is dirty (the set() method
         * or the setDirty() method has been explicitly called since the store creation,
         * or the last time setPristine() was called).
         * @return {?}
         */
        function () {
            return !this.store.isPristine;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Query.prototype, "config", {
        get: /**
         * @return {?}
         */
        function () {
            return this.constructor[queryConfigKey];
        },
        enumerable: true,
        configurable: true
    });
    return Query;
}());
/**
 * @template S
 */
export { Query };
if (false) {
    /**
     * Use only for internal plugins like Pagination - don't use this property *
     * @type {?}
     */
    Query.prototype.__store__;
    /** @type {?} */
    Query.prototype.store;
}
