// @datorama/akita/esm5/src/api/store-utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { isNumber } from '../internal/utils';
import { __stores__ } from './store';
import { applyTransaction } from './transaction';
/**
 * \@example
 *
 * this.peopleStore.update(id, person => ({
 *   guests: increment(person.guests)
 * }));
 * @param {?} value
 * @param {?=} params
 * @return {?}
 */
export function increment(value, params) {
    if (params === void 0) { params = { maxValue: undefined }; }
    if (isNumber(value) === false)
        return;
    if (params.maxValue && value === params.maxValue)
        return value;
    return value + 1;
}
/**
 * \@example
 *
 * this.peopleStore.update(id, person => ({
 *   guests: decrement(person.guests)
 * }));
 * @param {?} value
 * @param {?=} params
 * @return {?}
 */
export function decrement(value, params) {
    if (params === void 0) { params = { allowNegative: false }; }
    if (isNumber(value) === false)
        return;
    if (params.allowNegative === false && value === 0) {
        return value;
    }
    return value - 1;
}
/**
 * Generate random guid
 * @return {?}
 */
export function guid() {
    return 'xxxxxx4xyx'.replace(/[xy]/g, function (c) {
        /** @type {?} */
        var r = (Math.random() * 16) | 0;
        /** @type {?} */
        var v = c == 'x' ? r : (r & 0x3) | 0x8;
        return v.toString(16);
    });
}
/**
 * @record
 */
export function ResetStoresParams() { }
if (false) {
    /**
     *  By default the whole state is resetted, use this param to exclude stores that you don't want to reset.
     * @type {?}
     */
    ResetStoresParams.prototype.exclude;
}
/**
 *
 * @param {?=} options
 * @return {?}
 */
export function resetStores(options) {
    /** @type {?} */
    var defaults = {
        exclude: []
    };
    options = Object.assign({}, defaults, options);
    /** @type {?} */
    var stores = Object.keys(__stores__);
    applyTransaction(function () {
        var e_1, _a;
        try {
            for (var stores_1 = tslib_1.__values(stores), stores_1_1 = stores_1.next(); !stores_1_1.done; stores_1_1 = stores_1.next()) {
                var store = stores_1_1.value;
                /** @type {?} */
                var s = __stores__[store];
                if (!options.exclude) {
                    s.reset();
                }
                else {
                    if (options.exclude.indexOf(s.storeName) === -1) {
                        s.reset();
                    }
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (stores_1_1 && !stores_1_1.done && (_a = stores_1.return)) _a.call(stores_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    });
}
