// @datorama/akita/esm5/src/api/worker.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Observable } from 'rxjs';
/**
 * @param {?} object
 * @return {?}
 */
function serialize(object) {
    return JSON.stringify(object, function (key, value) {
        if (typeof value === 'function') {
            return value.toString();
        }
        return value;
    });
}
/**
 * @return {?}
 */
function createWorker() {
    /** @type {?} */
    var blob = new Blob([
        "self.onmessage = function(e) {\n\n      function deserialize(str) {\n        return JSON.parse(str || '', function(key, value) {\n          if (value &&\n            typeof value === \"string\" &&\n            value.substr(0, 8) == \"function\") {\n            var startBody = value.indexOf('{') + 1;\n            var endBody = value.lastIndexOf('}');\n            var startArgs = value.indexOf('(') + 1;\n            var endArgs = value.indexOf(')');\n            return new Function(value.substring(startArgs, endArgs), value.substring(startBody, endBody));\n          }\n          return value;\n        });\n      }\n\n      var deserialized = deserialize(e.data);\n      var mapped = deserialized.data.map(function(d) {\n        return deserialized.factory(d);\n      });\n\n      self.postMessage(mapped);\n    }"
    ], {
        type: 'text/javascript'
    });
    /** @type {?} */
    var url = URL.createObjectURL(blob);
    return new Worker(url);
}
/**
 * \@example
 *   const mockTodos = Array.from({length: 10000}, (_, x) => ({id: x}));
 *
 *   of(mockTodos).pipe(mapInWorker<Todo>(createTodo))
 *   .subscribe(res => console.log(res));
 * @template T
 * @param {?} factoryFn
 * @return {?}
 */
export function mapInWorker(factoryFn) {
    return function (source) {
        return new Observable(function (observer) {
            /** @type {?} */
            var worker = createWorker();
            worker.onmessage = function (e) {
                observer.next(e.data);
                observer.complete();
                worker.terminate();
            };
            worker.onerror = function (err) {
                observer.error(err);
                observer.complete();
                worker.terminate();
            };
            return source.subscribe(function (value) {
                /** @type {?} */
                var input = {
                    factory: factoryFn,
                    data: value
                };
                /** @type {?} */
                var serialized = serialize(input);
                worker.postMessage(serialized);
            });
        });
    };
}
