// @datorama/akita/esm5/src/internal/error.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { entityExists, isObject, toBoolean } from './utils';
var AkitaError = /** @class */ (function (_super) {
    tslib_1.__extends(AkitaError, _super);
    function AkitaError(message) {
        return _super.call(this, message) || this;
    }
    return AkitaError;
}(Error));
export { AkitaError };
var AkitaImmutabilityError = /** @class */ (function (_super) {
    tslib_1.__extends(AkitaImmutabilityError, _super);
    function AkitaImmutabilityError(storeName) {
        return _super.call(this, "The new state should be immutable. Make sure to return a new immutable state. \n store: \n " + storeName) || this;
    }
    return AkitaImmutabilityError;
}(AkitaError));
export { AkitaImmutabilityError };
var AkitaEntityNotExistsError = /** @class */ (function (_super) {
    tslib_1.__extends(AkitaEntityNotExistsError, _super);
    function AkitaEntityNotExistsError(id) {
        return _super.call(this, "Entity " + id + " does not exists") || this;
    }
    return AkitaEntityNotExistsError;
}(AkitaError));
export { AkitaEntityNotExistsError };
var AkitaNoActiveError = /** @class */ (function (_super) {
    tslib_1.__extends(AkitaNoActiveError, _super);
    function AkitaNoActiveError() {
        return _super.call(this, "Active is null/undefined") || this;
    }
    return AkitaNoActiveError;
}(AkitaError));
export { AkitaNoActiveError };
var AkitaInvalidEntityState = /** @class */ (function (_super) {
    tslib_1.__extends(AkitaInvalidEntityState, _super);
    function AkitaInvalidEntityState() {
        return _super.call(this, "Entity state is invalid") || this;
    }
    return AkitaInvalidEntityState;
}(AkitaError));
export { AkitaInvalidEntityState };
var AkitaUpdateIdKeyError = /** @class */ (function (_super) {
    tslib_1.__extends(AkitaUpdateIdKeyError, _super);
    function AkitaUpdateIdKeyError() {
        return _super.call(this, "Updating entity id is not permitted when updating many entities") || this;
    }
    return AkitaUpdateIdKeyError;
}(AkitaError));
export { AkitaUpdateIdKeyError };
/**
 * @param {?} id
 * @param {?} entities
 * @return {?}
 */
export function assertEntityExists(id, entities) {
    if (!entityExists(id, entities)) {
        throw new AkitaEntityNotExistsError(id);
    }
}
/**
 * @param {?} store
 * @return {?}
 */
export function assertActive(store) {
    if (!toBoolean(store.active)) {
        throw new AkitaNoActiveError();
    }
}
/**
 * @param {?} state
 * @return {?}
 */
export function assertEntityState(state) {
    /** @type {?} */
    var assertEntities = isObject(state);
    if (!assertEntities) {
        throw new AkitaInvalidEntityState();
    }
}
/**
 * @param {?} name
 * @param {?} className
 * @return {?}
 */
export function assertDecorator(name, className) {
    if (!name) {
        console.error("@StoreConfig({ name }) is missing in " + className);
    }
}
