// @datorama/akita/esm2015/src/plugins/dirty-check/entity-dirty-check-plugin.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { dirtyCheckDefaultParams, DirtyCheckPlugin, getNestedPath } from './dirty-check-plugin';
import { EntityCollectionPlugin } from '../entity-collection-plugin';
import { map, skip } from 'rxjs/operators';
/**
 * @template E, P
 */
export class EntityDirtyCheckPlugin extends EntityCollectionPlugin {
    /**
     * @param {?} query
     * @param {?=} params
     */
    constructor(query, params = {}) {
        super(query, params.entityIds);
        this.query = query;
        this.params = params;
        /**
         * @deprecated Use someDirty$ instead
         */
        this.isSomeDirty$ = this.query.select(state => state.entities).pipe(map((entities) => this.checkSomeDirty(entities)));
        this.someDirty$ = this.isSomeDirty$;
        this.someDirty = this.isSomeDirty;
        this.params = Object.assign({}, dirtyCheckDefaultParams, params);
        this.activate();
        this.selectIds()
            .pipe(skip(1))
            .subscribe(ids => {
            this.rebase(ids, { afterAdd: plugin => plugin.setHead() });
        });
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?=} ids
     * @return {THIS}
     */
    setHead(ids) {
        (/** @type {?} */ (this)).forEachId(ids, e => e.setHead());
        return (/** @type {?} */ (this));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    hasHead(id) {
        if (this.entities.has(id)) {
            /** @type {?} */
            const entity = this.getEntity(id);
            return entity.hasHead();
        }
        return false;
    }
    /**
     * @param {?=} ids
     * @param {?=} params
     * @return {?}
     */
    reset(ids, params = {}) {
        this.forEachId(ids, e => e.reset(params));
    }
    /**
     * @param {?} id
     * @param {?=} asObservable
     * @return {?}
     */
    isDirty(id, asObservable = true) {
        if (this.entities.has(id)) {
            /** @type {?} */
            const entity = this.getEntity(id);
            return asObservable ? entity.isDirty$ : entity.isDirty();
        }
        return false;
    }
    /**
     * @deprecated Use someDirty() instead
     * @return {?}
     */
    isSomeDirty() {
        /** @type {?} */
        const entities = this.query.getAll({ asObject: true });
        return this.checkSomeDirty(entities);
    }
    /**
     * @param {?} id
     * @param {?} path
     * @return {?}
     */
    isPathDirty(id, path) {
        if (this.entities.has(id)) {
            /** @type {?} */
            const head = ((/** @type {?} */ (this.getEntity(id)))).getHead();
            /** @type {?} */
            const current = this.query.getEntity(id);
            /** @type {?} */
            const currentPathValue = getNestedPath(current, path);
            /** @type {?} */
            const headPathValue = getNestedPath(head, path);
            return this.params.comparator(currentPathValue, headPathValue);
        }
        return null;
    }
    /**
     * @param {?=} ids
     * @return {?}
     */
    destroy(ids) {
        this.forEachId(ids, e => e.destroy());
    }
    /**
     * @param {?} id
     * @return {?}
     */
    instantiatePlugin(id) {
        return (/** @type {?} */ (new DirtyCheckPlugin(this.query, this.params, id)));
    }
    /**
     * @param {?} entities
     * @return {?}
     */
    checkSomeDirty(entities) {
        /** @type {?} */
        const entitiesIds = this.resolvedIds();
        for (const id of entitiesIds) {
            /** @type {?} */
            const dirty = this.params.comparator(((/** @type {?} */ (this.getEntity(id)))).getHead(), entities[id]);
            if (dirty) {
                return true;
            }
        }
        return false;
    }
}
if (false) {
    /**
     * @deprecated Use someDirty$ instead
     * @type {?}
     */
    EntityDirtyCheckPlugin.prototype.isSomeDirty$;
    /** @type {?} */
    EntityDirtyCheckPlugin.prototype.someDirty$;
    /** @type {?} */
    EntityDirtyCheckPlugin.prototype.someDirty;
    /** @type {?} */
    EntityDirtyCheckPlugin.prototype.query;
    /** @type {?} */
    EntityDirtyCheckPlugin.prototype.params;
}
