// @datorama/akita/esm2015/src/api/query.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { take } from 'rxjs/operators';
import { queryConfigKey } from './query-config';
/**
 * @template S
 */
export class Query {
    /**
     * @param {?} store
     */
    constructor(store) {
        this.store = store;
        this.__store__ = store;
    }
    /**
     * @template R
     * @param {?=} project
     * @return {?}
     */
    select(project) {
        /** @type {?} */
        let state = project ? project : state => state;
        return this.store._select(state);
    }
    /**
     * Select once and complete.
     * @template R
     * @param {?} project
     * @return {?}
     */
    selectOnce(project) {
        return this.select(project).pipe(take(1));
    }
    /**
     * Select the store's loading state.
     * @return {?}
     */
    selectLoading() {
        return this.select(state => ((/** @type {?} */ (state))).loading);
    }
    /**
     * Select the store's error state.
     * @return {?}
     */
    selectError() {
        return this.select(state => ((/** @type {?} */ (state))).error);
    }
    /**
     * Get the raw value of the store.
     * @return {?}
     */
    getSnapshot() {
        return this.store._value();
    }
    /**
     *  Returns whether the state of the store is pristine (the set() method hasn't been explicitly called
     *  since the store creation, or the last time setPristine()was called).
     * @return {?}
     */
    get isPristine() {
        return this.store.isPristine;
    }
    /**
     * Returns whether the state of the store is dirty (the set() method
     * or the setDirty() method has been explicitly called since the store creation,
     * or the last time setPristine() was called).
     * @return {?}
     */
    get isDirty() {
        return !this.store.isPristine;
    }
    /**
     * @return {?}
     */
    get config() {
        return this.constructor[queryConfigKey];
    }
}
if (false) {
    /**
     * Use only for internal plugins like Pagination - don't use this property *
     * @type {?}
     */
    Query.prototype.__store__;
    /** @type {?} */
    Query.prototype.store;
}
