// @datorama/akita/esm2015/src/api/store-utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { isNumber } from '../internal/utils';
import { __stores__ } from './store';
import { applyTransaction } from './transaction';
/**
 * \@example
 *
 * this.peopleStore.update(id, person => ({
 *   guests: increment(person.guests)
 * }));
 * @param {?} value
 * @param {?=} params
 * @return {?}
 */
export function increment(value, params = { maxValue: undefined }) {
    if (isNumber(value) === false)
        return;
    if (params.maxValue && value === params.maxValue)
        return value;
    return value + 1;
}
/**
 * \@example
 *
 * this.peopleStore.update(id, person => ({
 *   guests: decrement(person.guests)
 * }));
 * @param {?} value
 * @param {?=} params
 * @return {?}
 */
export function decrement(value, params = { allowNegative: false }) {
    if (isNumber(value) === false)
        return;
    if (params.allowNegative === false && value === 0) {
        return value;
    }
    return value - 1;
}
/**
 * Generate random guid
 * @return {?}
 */
export function guid() {
    return 'xxxxxx4xyx'.replace(/[xy]/g, c => {
        /** @type {?} */
        const r = (Math.random() * 16) | 0;
        /** @type {?} */
        const v = c == 'x' ? r : (r & 0x3) | 0x8;
        return v.toString(16);
    });
}
/**
 * @record
 */
export function ResetStoresParams() { }
if (false) {
    /**
     *  By default the whole state is resetted, use this param to exclude stores that you don't want to reset.
     * @type {?}
     */
    ResetStoresParams.prototype.exclude;
}
/**
 *
 * @param {?=} options
 * @return {?}
 */
export function resetStores(options) {
    /** @type {?} */
    const defaults = {
        exclude: []
    };
    options = Object.assign({}, defaults, options);
    /** @type {?} */
    const stores = Object.keys(__stores__);
    applyTransaction(() => {
        for (const store of stores) {
            /** @type {?} */
            const s = __stores__[store];
            if (!options.exclude) {
                s.reset();
            }
            else {
                if (options.exclude.indexOf(s.storeName) === -1) {
                    s.reset();
                }
            }
        }
    });
}
